applybn/feature_extraction/bn_feature_extractor.py

```diff
@@ -1,312 +1,323 @@
-
+        self.known_structure = known_structure
+        self.bn = None
+        self.variables: Optional[List[str]] = None
+        self.encoder = preprocessing.LabelEncoder()
+        self.discretizer = preprocessing.KBinsDiscretizer(
+            n_bins=5,
+            encode='ordinal',
+            strategy='kmeans',
+            subsample=None
+        )
+        self.black_list = black_list
+        self.preprocessor = pp.Preprocessor([('encoder', self.encoder), ('discretizer', self.discretizer)])
+        self.nodes_dict = {} # Dictionary to store nodes of the Bayesian Network
+        self.target_name = ''
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
         """
-        Initializes the BNFeatureGenerator.
+        Fits the BNFeatureGenerator to the data.
+
+        This involves:
+        1.  Adding the target variable (if provided) to the input data.
+        2.  Encoding categorical columns.
+        3.  Discretizing continuous columns.
+        4.  Creating a Bayesian Network based on the data types.
+        5.  Learning the structure of the Bayesian Network (if known_structure is not provided).
+        6.  Fitting the parameters of the Bayesian Network.
 
         Args:
-            known_structure (Optional[List[Tuple[str, str]]]): A list of edges
-                representing the known structure of the Bayesian Network.
+            X (pd.DataFrame): The input data.
+            y (Optional[pd.Series]): The target variable. If provided, it will be added to the input data and
+                treated as a node in the Bayesian Network.
+
+        Returns:
+            self: The fitted BNFeatureGenerator object.
         """
-        self.known_structure = known_structure
-        self.bn: Optional[BayesianNetwork] = None
-        self.variables: Optional[List[str]] = None
-        self.num_classes: Optional[int] = None
-        self.random_seed = random_seed
+        if y is not(None):
+            X = pd.concat([X, y], axis = 1).reset_index(drop=True)
+            self.target_name = y.name
+        cat_columns = X.select_dtypes(include=["object", "category"]).columns.tolist()
+
+        if cat_columns:
+            X, cat_encoder = code_categories(X, method="label", columns=cat_columns)
+        X = pd.DataFrame(X)
+        discretized_data, est = self.preprocessor.apply(X)
+        discretized_data = pd.DataFrame(discretized_data, columns=X.columns)
+        info = self.preprocessor.info # Get information about the data types after preprocessing
+        disc_columns = 0
+        cont_columns = 0
+        get_nodes_type = info['types'] # Get types of nodes
+        print(get_nodes_type)
+
+        for key, value in get_nodes_type.items(): # Iterate through node types and count discrete and continuous columns
+            if value == 'disc' or value == 'disc_num':
+                disc_columns += 1
+            else:
+                cont_columns += 1
+        disc_columns+=len(cat_columns)
```

Comment:
Это можно сделать проще:

```suggestion
        # Get types of nodes
        get_nodes_type = info['types']
        print(get_nodes_type)
        
        # Check for discrete and continuous columns
        values = get_nodes_type.values()
        has_discrete = any(value in ['disc', 'disc_num'] for value in values)
        has_continuous = any(value not in ['disc', 'disc_num'] for value in values)
        
        # If we have categorical columns, we have discrete data
        if cat_columns:
            has_discrete = True
```